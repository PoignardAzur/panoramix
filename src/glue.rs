//! Bunch of types mostly stolen from Crochet, to use as glue with Druid

#![allow(unused)]

use druid::Data;
use std::collections::HashMap;
use std::sync::Arc;

pub use druid::debug_state::DebugState;

pub struct GlobalEventCx<'a> {
    pub app_data: &'a mut DruidAppData,
}

impl<'a> GlobalEventCx<'a> {
    pub fn new(app_data: &'a mut DruidAppData) -> Self {
        Self { app_data }
    }
}

/// The type we use for app data for Druid integration.
///
/// Currently this is action queues.
///
/// It should probably be a vec of actions, but we can refine
/// later. For button clicks it doesn't matter.
#[derive(Clone, Data, Default)]
pub struct DruidAppData(Arc<HashMap<Id, Action>>);

/// Actions that can be produced by widgets,
#[derive(Clone)]
pub enum Action {
    Clicked,
    FutureResolved,
    TextChanged(String),
}

impl DruidAppData {
    pub(crate) fn queue_action(&mut self, id: Id, action: Action) {
        Arc::make_mut(&mut self.0).insert(id, action);
    }

    pub(crate) fn dequeue_action(&mut self, id: Id) -> Option<Action> {
        if self.0.contains_key(&id) {
            Arc::make_mut(&mut self.0).remove(&id)
        } else {
            None
        }
    }

    /// Report whether the id has a non-empty action queue.
    pub(crate) fn has_action(&self, id: Id) -> bool {
        self.0.contains_key(&id)
    }
}

// ---

use std::sync::atomic::{AtomicUsize, Ordering};

static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);

/// An identifier for an element.
///
/// It's a bit heavy-handed to have this id as well as widget
/// id in Druid; likely the two concepts should be unified. But
#[derive(Clone, Copy, Debug, PartialOrd, PartialEq, Ord, Eq, Hash)]
pub struct Id(usize);

impl Id {
    /// Allocate a new unique id.
    pub fn new() -> Id {
        Id(ID_COUNTER.fetch_add(1, Ordering::Relaxed))
    }
}

// ---

use tracing_subscriber::prelude::*;
use tracing_subscriber::{fmt, EnvFilter};
use tracing_unwrap::ResultExt;

pub fn init_tracing() {
    let fmt_layer = fmt::layer().with_target(true);
    let filter_layer = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("warn"))
        .unwrap_or_log();

    tracing_subscriber::registry()
        .with(filter_layer)
        .with(fmt_layer)
        .init();
}
