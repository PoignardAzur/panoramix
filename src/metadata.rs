use derivative::Derivative;

/// Placeholder type for elements that don't raise events.
///
/// Equivalent to `!`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum NoEvent {}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
pub struct NoState;

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Copy(bound = ""),
    Default(bound = ""),
    PartialEq(bound = ""),
    Eq(bound = ""),
    Hash(bound = "")
)]
pub struct Metadata<ComponentEvent, ComponentState> {
    _marker: std::marker::PhantomData<(ComponentEvent, ComponentState)>,
}

impl<ComponentEvent, ComponentState> Metadata<ComponentEvent, ComponentState> {
    pub fn new() -> Self {
        Default::default()
    }
}
