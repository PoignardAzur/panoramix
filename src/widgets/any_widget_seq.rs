use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

use std::any::Any;

pub trait AnyWidgetSeq: Any {
    fn as_any(&self) -> &dyn Any;
    fn as_mut_any(&mut self) -> &mut dyn Any;
    fn any_widgets(&self) -> Vec<&dyn FlexWidget>;
    fn any_widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget>;
}

impl<T> AnyWidgetSeq for T
where
    T: WidgetSequence + 'static,
{
    fn as_any(&self) -> &dyn Any {
        self
    }
    fn as_mut_any(&mut self) -> &mut dyn Any {
        self
    }

    fn any_widgets(&self) -> Vec<&dyn FlexWidget> {
        self.widgets()
    }

    fn any_widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.widgets_mut()
    }
}

// -

pub struct WidgetSeqBox {
    pub value: Box<dyn AnyWidgetSeq>,
}

impl WidgetSequence for WidgetSeqBox {
    fn widgets(&self) -> Vec<&dyn FlexWidget> {
        self.value.any_widgets()
    }

    fn widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.value.any_widgets_mut()
    }
}
