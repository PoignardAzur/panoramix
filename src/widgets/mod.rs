//! Wrapper types around druid widgets.

pub mod any_widget_seq;
pub mod button_widget;
pub mod checkbox_widget;
pub mod empty_sequence;
pub mod flex_widget;
pub mod optional_widget;
pub mod single_widget;
pub mod textbox_widget;
pub mod widget_list;
pub mod widget_tuple;

pub use any_widget_seq::*;
pub use button_widget::*;
pub use checkbox_widget::*;
pub use empty_sequence::*;
pub use flex_widget::*;
pub use optional_widget::*;
pub use single_widget::*;
pub use textbox_widget::*;
pub use widget_list::*;
pub use widget_tuple::*;
