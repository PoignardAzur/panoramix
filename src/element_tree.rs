use crate::glue::{DruidAppData, GlobalEventCx};
use crate::widget_sequence::WidgetSequence;

use druid::{Env, EventCtx};
use std::any::Any;
use std::fmt::Debug;

/// Context type passed to all components when building them.
pub struct CompCtx {
    pub(crate) local_state: Box<dyn Any>,
}

impl CompCtx {
    /// Returns the local state of the current component instance.
    ///
    /// Panics if the generic type doesn't match the component's local state type.
    pub fn use_local_state<T: 'static>(&self) -> &T {
        self.local_state.downcast_ref::<T>().unwrap()
    }

    // TODO - add methods
    // use_lifecycle
    // get_vdom_context
}

/// Context required by [`VirtualDom::reconcile`]
pub struct ReconcileCtx<'a, 'b, 'c, 'd, 'e> {
    pub event_ctx: &'a mut EventCtx<'d, 'e>,
    pub data: &'b mut DruidAppData,
    pub env: &'c Env,
}

/// The trait implemented by all GUI elements.
///
/// Every type you use to explicitly create a GUI in Panoramix ([`Button`](crate::elements::Button), [`TextBox`](crate::elements::TextBox), any user-made component) implements Element. You usually don't need to worry about this trait unless you want to implement your own custom element.
///
/// For helper methods that can be called on all elements, see [`ElementExt`].
///
/// ## Note about template resolution
///
/// This trait is parameterized on two template types: `CpEvent` and `CpState`, which represent the event and local-state type of the parent component an element is built in. They are supposed to flow "inwards" with type inference, starting from the `-> impl Element<MyEvent, MyState>` return type of your function.
///
/// To give a concrete example:
///
/// ```rust
/// # use panoramix::{component, CompCtx, Column, Element, ElementExt};
/// # use panoramix::elements::{ButtonClick, Button, Label};
/// # type BuyItem = ButtonClick;
/// #
/// #[component]
/// fn StoreItem(ctx: &CompCtx, item_name: String) -> impl Element<BuyItem, u32> {
///     let item_count = ctx.use_local_state::<u32>();
///     Column!(
///         Label::new(format!("Item: {} x{}", item_name, item_count)),
///         Button::new("+")
///             .on_click(|item_count, _| {
///                 *item_count += 1;
///             }),
///         Button::new("Buy")
///             .bubble_up::<BuyItem>()
///     )
/// }
/// ```
///
/// In this example, because the return type is declared to be `-> impl Element<BuyItem, u32>`, all elements that are returned (Label, Button, Column) will be transitively inferred to implement `Element<BuyItem, u32>`.
///
/// The flip side of this is that constructing an element and not returning it (eg doing `let x = Button::new("...");` and then not using `x`) will lead to a compile error, because the compiler can't infer what `CpEvent` and `CpState` should be.
///
pub trait Element<CpEvent = NoEvent, CpState = ()>: Debug {
    /// The type of events this element can raise.
    ///
    /// This is the type that [`ElementExt::on`], [`ElementExt::map_event`] and [`ElementExt::bubble_up`] can take. It's different from the `CpEvent` generic parameter, which is the event the parent component emits.
    ///
    /// In the `StoreItem` example, the `Event` type of buttons is `ButtonClick`, and their `CpEvent` parameter is `BuyItem`.
    type Event;

    type AggregateChildrenState: Clone + Default + Debug + PartialEq;
    type BuildOutput: VirtualDom<
        CpEvent,
        CpState,
        Event = Self::Event,
        AggregateChildrenState = Self::AggregateChildrenState,
    >;

    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState);
}

// TODO - Include documentation about what a Virtual DOM is and where the name comes from.
pub trait VirtualDom<CpEvent, CpState>: Debug {
    type AggregateChildrenState: Clone + Default + Debug + PartialEq;
    type TargetWidgetSeq: WidgetSequence;

    type Event;

    // update_value is intended to enable memoize-style HOC
    // where instead of returning a vdom node, it returns
    // something along the lines of struct KeepEverythingAsItWas()
    // Ugh. I'm not explaining this well.
    fn update_value(&mut self, other: Self)
    where
        Self: Sized,
    {
        *self = other;
    }

    fn init_tree(&self) -> Self::TargetWidgetSeq;

    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    );

    // TODO - Rename methods
    #[allow(unused_variables)]
    fn process_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<CpEvent> {
        None
    }

    #[allow(unused_variables)]
    fn process_local_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        None
    }
}

/// Placeholder type for elements that don't raise events.
///
/// Equivalent to `!`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum NoEvent {}

// Used in unit tests
#[allow(dead_code)]
pub(crate) fn assign_empty_state_type(_elem: &impl Element<NoEvent, ()>) {}

#[allow(dead_code)]
pub(crate) fn assign_state_type<CpEvent, CpState, Elem: Element<CpEvent, CpState>>(_elem: &Elem) {}

use crate::elements::with_event::{ParentEvent, WithBubbleEvent, WithCallbackEvent, WithMapEvent};

/// Helper methods that can be called on all elements.
pub trait ElementExt<CpEvent, CpState>: Element<CpEvent, CpState> + Sized {
    fn on<EventParam, Cb: Fn(&mut CpState, EventParam)>(
        self,
        callback: Cb,
    ) -> WithCallbackEvent<CpEvent, CpState, EventParam, Self, Cb>
    where
        Self::Event: ParentEvent<EventParam>,
    {
        WithCallbackEvent {
            element: self,
            callback,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
        }
    }

    fn map_event<EventParam, EventReturn, Cb: Fn(&mut CpState, EventParam) -> Option<EventReturn>>(
        self,
        callback: Cb,
    ) -> WithMapEvent<CpEvent, CpState, EventParam, EventReturn, Self, Cb>
    where
        Self::Event: ParentEvent<EventParam>,
        CpEvent: ParentEvent<EventReturn>,
    {
        WithMapEvent {
            element: self,
            callback,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
            _comp_return: Default::default(),
        }
    }

    fn bubble_up<Event>(self) -> WithBubbleEvent<CpEvent, CpState, Event, Self>
    where
        Self::Event: ParentEvent<Event>,
        CpEvent: ParentEvent<Event>,
    {
        WithBubbleEvent {
            element: self,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
        }
    }
}

impl<CpEvent, CpState, ET: Element<CpEvent, CpState>> ElementExt<CpEvent, CpState> for ET {}
