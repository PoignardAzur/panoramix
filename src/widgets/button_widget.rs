use crate::flex::FlexParams;
use crate::glue::{Action, DruidAppData, WidgetId};
use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

use crate::glue::DebugState;
use druid::kurbo::{Point, Rect, Size};
use druid::widget::{Button, Click, ControllerHost, IdentityWrapper};
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, UpdateCtx,
    Widget, WidgetExt, WidgetPod,
};

pub struct ButtonWidget {
    pub pod: WidgetPod<
        DruidAppData,
        IdentityWrapper<ControllerHost<Button<DruidAppData>, Click<DruidAppData>>>,
    >,
    pub flex: FlexParams,
    id: WidgetId,
}

impl ButtonWidget {
    pub fn new(text: String, flex: FlexParams, id: WidgetId) -> Self {
        let button = Button::new(text)
            .on_click(move |_, data: &mut DruidAppData, _| data.queue_action(id, Action::Clicked))
            .with_id(id);

        ButtonWidget {
            pod: WidgetPod::new(button),
            flex,
            id,
        }
    }

    pub fn id(&self) -> WidgetId {
        self.id
    }
}

impl FlexWidget for ButtonWidget {
    fn flex_params(&self) -> FlexParams {
        self.flex
    }

    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut DruidAppData, env: &Env) {
        self.pod.event(ctx, event, data, env);
    }

    fn lifecycle(
        &mut self,
        ctx: &mut LifeCycleCtx,
        event: &LifeCycle,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.pod.lifecycle(ctx, event, data, env);
    }

    fn update(
        &mut self,
        ctx: &mut UpdateCtx,
        _old_data: &DruidAppData,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.pod.update(ctx, data, env);
    }

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        bc: &BoxConstraints,
        data: &DruidAppData,
        env: &Env,
    ) -> Size {
        self.pod.layout(ctx, bc, data, env)
    }

    fn paint_rect(&self) -> Rect {
        self.pod.paint_rect()
    }

    fn set_origin(&mut self, ctx: &mut LayoutCtx, data: &DruidAppData, env: &Env, origin: Point) {
        self.pod.set_origin(ctx, data, env, origin)
    }

    fn layout_rect(&self) -> Rect {
        self.pod.layout_rect()
    }

    fn baseline_offset(&self) -> f64 {
        self.pod.baseline_offset()
    }

    fn paint(&mut self, ctx: &mut PaintCtx, data: &DruidAppData, env: &Env) {
        self.pod.paint(ctx, data, env);
    }

    fn debug_state(&self, data: &DruidAppData) -> DebugState {
        self.pod.widget().debug_state(data)
    }
}

impl WidgetSequence for ButtonWidget {
    fn widgets(&self) -> Vec<&dyn FlexWidget> {
        vec![self]
    }

    fn widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget> {
        vec![self]
    }
}
