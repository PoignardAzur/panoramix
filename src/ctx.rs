use crate::glue::DruidAppData;
use crate::metadata::{Metadata, NoState};
use druid::{Env, EventCtx};
use std::any::{type_name, Any, TypeId};

/// Context type passed to all components when building them.
pub struct CompCtx<'a> {
    // Yeah, we're using a cell, sorry. It's very local, though.
    pub(crate) called_use_metadata: std::cell::Cell<bool>,
    pub(crate) local_state: &'a dyn Any,
}

impl<'a> CompCtx<'a> {
    pub fn use_metadata<ComponentEvent: 'static, ComponentState: 'static>(
        &self,
    ) -> Metadata<ComponentEvent, ComponentState> {
        if self.called_use_metadata.get() {
            panic!("error: 'use_metadata' can only be called once per component")
        }
        self.called_use_metadata.set(true);
        Default::default()
    }

    /// Returns the local state of the current component instance.
    ///
    /// Panics if the generic type doesn't match the component's local state type.
    pub fn get_local_state<ComponentEvent: 'static, ComponentState: 'static>(
        &self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &'a ComponentState {
        #![allow(unused_variables)]
        if (*self.local_state).type_id() == TypeId::of::<NoState>() {
            panic!("error: 'get_local_state' cannot be called for a component whose root element isn't ComponentOutput")
        }
        self.local_state
            .downcast_ref::<ComponentState>()
            .expect(&format!(
            "internal type error: get_local_state expected {:?} ({}), parent component gave {:?}",
            TypeId::of::<ComponentState>(),
            type_name::<ComponentState>(),
            (*self.local_state).type_id(),
        ))
    }

    // TODO - add methods
    // use_lifecycle
    // get_vdom_context
}

/// Context required by [`VirtualDom::reconcile`]
pub struct ReconcileCtx<'a, 'b, 'c, 'd, 'e> {
    pub event_ctx: &'a mut EventCtx<'d, 'e>,
    pub data: &'b mut DruidAppData,
    pub env: &'c Env,
}

pub struct ProcessEventCtx<'e, 's> {
    pub event_queue: &'e mut dyn Any,
    pub state: &'s mut dyn Any,
}

impl<'e, 's> ProcessEventCtx<'e, 's> {
    pub fn event_queue<ComponentEvent: 'static, ComponentState: 'static>(
        &mut self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &mut Vec<ComponentEvent> {
        #![allow(unused_variables)]
        let type_id = (*self.event_queue).type_id();
        self.event_queue
            .downcast_mut::<Vec<ComponentEvent>>()
            .expect(&format!(
                "internal type error: event handler expected {:?} ({}), parent component gave {:?}",
                TypeId::of::<Vec<ComponentEvent>>(),
                type_name::<Vec<ComponentEvent>>(),
                type_id,
            ))
    }

    pub fn state<ComponentEvent: 'static, ComponentState: 'static>(
        &mut self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &mut ComponentState {
        #![allow(unused_variables)]
        let type_id = (*self.event_queue).type_id();
        self.state.downcast_mut::<ComponentState>().expect(&format!(
            "internal type error: event handler expected {:?} ({}), parent component gave {:?}",
            TypeId::of::<ComponentState>(),
            type_name::<ComponentState>(),
            type_id,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::metadata::NoEvent;
    use test_env_log::test;

    #[derive(Debug, PartialEq, Eq)]
    struct MyEvent(i32);

    #[test]
    fn event_queue() {
        let md: Metadata<MyEvent, NoState> = Default::default();

        let mut event_queue = Vec::<MyEvent>::new();
        let mut ctx = ProcessEventCtx {
            event_queue: &mut event_queue,
            state: &mut NoState,
        };
        ctx.event_queue(md).push(MyEvent(42));

        assert_eq!(event_queue, vec![MyEvent(42)],);
    }

    #[test]
    fn state() {
        let md: Metadata<NoEvent, i64> = Default::default();

        let mut state = 12345_i64;
        let mut ctx = ProcessEventCtx {
            event_queue: &mut Vec::<NoEvent>::new(),
            state: &mut state,
        };

        assert_eq!(*ctx.state(md), 12345_i64,);

        *ctx.state(md) = 123_i64;
        assert_eq!(state, 123_i64,);
    }
}
