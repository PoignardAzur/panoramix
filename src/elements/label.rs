use crate::element_tree::{Element, NoEvent, VirtualDom};
use crate::flex::FlexParams;
use crate::glue::DruidAppData;
use crate::widgets::SingleWidget;

use crate::element_tree::ReconcileCtx;
use druid::widget as druid_w;

use derivative::Derivative;
use tracing::instrument;

/// A text label.
///
/// ## Events
///
/// Doesn't emit events.
#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct Label<CpEvent = NoEvent, CpState = ()> {
    pub text: String,
    pub flex: FlexParams,
    #[derivative(Debug = "ignore")]
    pub _markers: std::marker::PhantomData<(CpEvent, CpState)>,
}

#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct LabelData<CpEvent = NoEvent, CpState = ()> {
    pub text: String,
    pub flex: FlexParams,
    #[derivative(Debug = "ignore")]
    pub _markers: std::marker::PhantomData<(CpEvent, CpState)>,
}

//
// --- IMPLS

impl<CpEvent, CpState> Label<CpEvent, CpState> {
    /// Build a text label.
    pub fn new(text: impl Into<String>) -> Label<CpEvent, CpState> {
        Label {
            text: text.into(),
            flex: FlexParams {
                flex: 1.0,
                alignment: None,
            },
            _markers: Default::default(),
        }
    }

    /// Change the way the label's size is calculated
    pub fn with_flex_params(self, flex_params: FlexParams) -> Self {
        Label {
            flex: flex_params,
            ..self
        }
    }

    // TODO
    /// Used for unit tests.
    pub fn with_mock_state(self) -> super::WithMockState<Self, CpEvent, CpState> {
        super::WithMockState::new(self)
    }
}

impl<CpEvent, CpState> LabelData<CpEvent, CpState> {
    pub fn new(text: impl Into<String>) -> LabelData<CpEvent, CpState> {
        LabelData {
            text: text.into(),
            flex: FlexParams {
                flex: 1.0,
                alignment: None,
            },
            _markers: Default::default(),
        }
    }

    pub fn with_mock_state(self) -> super::WithMockStateData<Self, CpEvent, CpState> {
        super::WithMockStateData::new(self)
    }
}

impl<CpEvent, CpState> Element<CpEvent, CpState> for Label<CpEvent, CpState> {
    type Event = NoEvent;
    type ComponentState = crate::element_tree::NoState;
    type AggregateChildrenState = ();
    type BuildOutput = LabelData<CpEvent, CpState>;

    #[instrument(name = "Label", skip(self, _prev_state))]
    fn build(self, _prev_state: ()) -> (LabelData<CpEvent, CpState>, ()) {
        (
            LabelData {
                text: self.text,
                flex: self.flex,
                _markers: Default::default(),
            },
            (),
        )
    }
}

impl<CpEvent, CpState> VirtualDom<CpEvent, CpState> for LabelData<CpEvent, CpState> {
    type Event = NoEvent;
    type AggregateChildrenState = ();
    type TargetWidgetSeq = SingleWidget<druid_w::Label<DruidAppData>>;

    #[instrument(name = "Label", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        let label = druid_w::Label::new(self.text.clone());
        SingleWidget::new(label, self.flex)
    }

    #[instrument(name = "Label", skip(self, other, widget, ctx))]
    fn reconcile(&self, other: &Self, widget: &mut Self::TargetWidgetSeq, ctx: &mut ReconcileCtx) {
        if self.text != other.text {
            widget.pod.widget_mut().set_text(self.text.clone());
            widget.request_druid_update(ctx);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::element_tree::assign_empty_state_type;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn new_label() {
        let label = Label::new("Hello");
        let (label_data, ()) = label.clone().build(());

        assert_debug_snapshot!(label);
        assert_debug_snapshot!(label_data);

        assert_eq!(label_data, LabelData::new("Hello"));

        assign_empty_state_type(&label);
    }

    #[test]
    fn test_label_widget() {
        use crate::test_harness::Harness;

        let label = Label::new("Hello");

        Harness::run_test_window(label, |harness| {
            let label_state = harness.get_root_debug_state();
            assert_debug_snapshot!(label_state);

            let new_label = Label::new("World");
            harness.update_root_element(new_label);

            let label_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(label_state_2);
        });
    }
}
