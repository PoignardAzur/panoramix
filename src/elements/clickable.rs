use crate::ctx::ReconcileCtx;
use crate::element_tree::{Element, ElementExt, VirtualDom};
use crate::glue::{Action, GlobalEventCx, WidgetId};
use crate::metadata::{Metadata, NoState};
use crate::widgets::ClickableWidget;

use tracing::{instrument, trace};

// TODO - doc
#[derive(Clone, Debug, Default, PartialEq)]
pub struct Clickable<Child: Element> {
    pub child: Child,
    pub reserved_widget_id: Option<WidgetId>,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ClickableData<Child: VirtualDom> {
    pub child: Child,
    pub reserved_widget_id: Option<WidgetId>,
}

/// Event emitted when a [Clickable] is clicked.
///
/// Note: Might hold data like "mouse position" or "widget id" future versions.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct ClickEvent;

//
// --- IMPLS

impl<Child: Element> Clickable<Child> {
    pub fn new(child: Child) -> Self {
        Clickable {
            child,
            reserved_widget_id: None,
        }
    }

    /// For unit tests only.
    ///
    /// The widget created by this element always has the same id. If two widgets are created
    /// with the same id (for instance, because the same widget is returned twice because of
    /// a copy-paste error), impredictable behavior may ensue.
    pub fn with_reserved_id(self, widget_id: WidgetId) -> Self {
        Clickable {
            reserved_widget_id: Some(widget_id),
            ..self
        }
    }

    /// Provide a closure to be called when the child widget is clicked.
    pub fn on_click<ComponentEvent: 'static, ComponentState: 'static>(
        self,
        md: Metadata<ComponentEvent, ComponentState>,
        callback: impl Fn(&mut ComponentState, ClickEvent) + Clone + 'static,
    ) -> impl Element {
        self.on(md, callback)
    }
}

impl<Child: Element> Element for Clickable<Child> {
    type Event = ClickEvent;

    type ComponentState = NoState;
    type AggregateChildrenState = Child::AggregateChildrenState;
    type BuildOutput = ClickableData<Child::BuildOutput>;

    #[instrument(name = "Clickable", skip(self, prev_state))]
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (
        ClickableData<Child::BuildOutput>,
        Self::AggregateChildrenState,
    ) {
        let (element, child_state) = self.child.build(prev_state);
        (
            ClickableData {
                child: element,
                reserved_widget_id: self.reserved_widget_id,
            },
            child_state,
        )
    }
}

impl<Child: VirtualDom> VirtualDom for ClickableData<Child> {
    type Event = ClickEvent;
    type AggregateChildrenState = Child::AggregateChildrenState;
    type TargetWidgetSeq = ClickableWidget<Child::TargetWidgetSeq>;

    #[instrument(name = "Clickable", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        let id = self.reserved_widget_id.unwrap_or_else(WidgetId::next);
        ClickableWidget::new(self.child.init_tree(), id)
    }

    #[instrument(name = "Clickable", skip(self, prev_value, widget_seq, ctx))]
    fn reconcile(
        &self,
        prev_value: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        self.child
            .reconcile(&prev_value.child, widget_seq.children_mut(), ctx);
    }

    #[instrument(name = "Clickable", skip(self, _children_state, widget, cx))]
    fn process_local_event(
        &self,
        _children_state: &mut Self::AggregateChildrenState,
        widget: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<ClickEvent> {
        // FIXME - Rework event dispatching
        let id = widget.id();
        if let Some(Action::Clicked) = cx.app_data.dequeue_action(id) {
            trace!("Processed click");
            Some(ClickEvent)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::elements::label::Label;
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn new_clickable() {
        let clickable = Clickable::new(Label::new("Hello"));

        let (clickable_data, ()) = clickable.clone().build(());

        assert_debug_snapshot!(clickable);
        assert_debug_snapshot!(clickable_data);
    }

    #[test]
    fn clickable_widget() {
        let clickable = Clickable::new(Label::new("Hello"));

        Harness::run_test_window(clickable, |_harness| {
            // FIXME
            //let clickable_state = harness.get_root_debug_state();
            //assert_debug_snapshot!(clickable_state);

            // FIXME - Test reconcile() method with child
        });
    }

    #[test]
    fn clickable_press() {
        use crate::elements::event_logger::EventLogger;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let clickable_id = WidgetId::reserved(1);
        let clickable = EventLogger::new(
            event_sender,
            Clickable::new(Label::new("Hello")).with_reserved_id(clickable_id),
        );

        Harness::run_test_window(clickable, |harness| {
            // FIXME
            // assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(clickable_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);
        });
    }
}
