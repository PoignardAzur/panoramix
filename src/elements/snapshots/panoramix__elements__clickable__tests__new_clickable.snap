---
source: src/elements/clickable.rs
expression: clickable

---
Clickable {
    child: Label {
        text: "Hello",
        flex: FlexParams {
            flex: None,
            alignment: None,
        },
    },
    reserved_widget_id: None,
}
